### PR TITLE
Enrich Docs with Usage

### PR DESCRIPTION
**Pull request by echarles**
This PR enriches the docs with explanations on how kernel_mgmt can be used as standalone or in jupyter_server.

(it also update the env to support markdown)

### CODE REVIEW

**Comment by echarles** [comment]
Thx for the review @kevin-bates. I will inject your comments + draw the interactions in the sequence diagram and will push all that (my) tomorrow.

**Comment by takluyver** [comment]
I would prefer to stick with rst for these docs. I know Markdown is more popular, and it's fine for things like Github comments, but the crosslinking Sphinx can do with rst roles is vastly superior. I also think using the same format across the docs is a good idea.

**Comment by echarles** [comment]
@takluyver Sure, I will upgrade my rst skills and convert tomorrow the markdows. I started with rst but came into sphynx confusions with the links (should be able to overcome that tomorrow).

File: docs/server.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Use with Jupyter Server
+
+This page describes the way [Jupyter Server](https://github.com/jupyter/jupyter_server) uses the Kernel Management module.
+
+![](./images/kernel_mgmt_server.svg "")
+
+On the Jupyter Server side, each WEB handlers (located in services/kernels/handlers: MainKernelHandler, KernelHandler, KernelActionHandler, ZMQChannelHandler) have:
+
+- A kernel_manager. The default manager is `jupyter_server.services.kernels.kernelmanager.MappingKernelManager`.
+- A kernel_finder. The default is `jupyter_kernel_mgmt.discovery.KernelFinder` which is imported from the jupyter_kernel_mgmt library.
```
[the remark this answers is not in the document]
Comment (echarles) [reply docs/server.md]: Done.

File: docs/standalone.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# Standalone Usage
+
+The aim of the Kernel Management is to be integrated in larger application such as [Jupyter Server](./server).
+
+Although, it can be used a standalone module to, for example, launch a [Kernel Finder](./kernel_finder) from the command line and get a list of [Kernel Specs](./kernel_finder.html#kernel-specifications).
+
+```python
+# !!! This is failing for now...
+# Follow https://github.com/takluyver/jupyter_kernel_mgmt/issues/27 for a fix.
```
[the remark this answers is not in the document]
Comment (echarles) [reply docs/standalone.md]: Great and blazing fast fix. Thx @kevin-bates

**Comment by echarles** [comment]
@kevin-bates @takluyver Thx for your reviews. I have pushed an update that take them into account (still need to work on the sequence diagram, will do that tomorrow for final review).

File: docs/images/kernel_mgmt_server_sequence.svg [hunk 3]
```diff
@@ -0,0 +1,1933 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
[the remark this answers is not in the document]
Comment (echarles) [reply docs/images/kernel_mgmt_server_sequence.svg]: Thx. need to working more on the sequence diagram.

[the remark this answers is not in the document]
Comment (kevin-bates) [reply docs/images/kernel_mgmt_server_sequence.svg]: The `/api/kernelspecs` call [hits KernelFinder directly](https://github.com/kevin-bates/jupyter_server/blob/jupyter-kernel-mgmt/jupyter_server/services/kernelspecs/handlers.py#L66) from the kernelspec handler.

File: docs/server.rst [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+.. code-block:: python
+
+  entrypoints:
+    jupyter_kernel_mgmt.kernel_type_providers' : [
+      'spec = jupyter_kernel_mgmt.discovery:KernelSpecProvider',
+      'pyimport = jupyter_kernel_mgmt.discovery:IPykernelProvider',
+    ]
+
+External Providers can register their own entypoints, e.g the `kubernetes_kernel_provider <https://github.com/gateway-experiments/kubernetes_kernel_provider>`_ which extends the `remote_kernel_provider <https://github.com/gateway-experiments/remote_kernel_provider>`_.
```
[the remark this answers is not in the document]
Comment (echarles) [reply docs/server.rst]: updated doc and text
@takluyver ping me if ok to add jupyter_ssh_kernels. Thx

File: docs/standalone.rst [hunk 5]
```diff
@@ -0,0 +1,14 @@
+.. _standalone:
+
+================
+Standalone Usage
+================
```
[the remark this answers is not in the document]
Comment (echarles) [reply docs/standalone.rst]: Added paragraphs for this.

**Comment by echarles** [comment]
@kevin-bates just pushed your comments + a better sequence diagram

File: docs/server.rst [hunk 6]
```diff
@@ -10,28 +10,32 @@ This page describes the way `Jupyter Server <https://github.com/jupyter/jupyter_
...
 
-All instances of MappingKernelManager have a ``KernelFinder`` field.
+A single instance of MappingKernelManager is shared across all other objects (singleton pattern).
+The MappingKernelManager instance has a ``KernelFinder`` field.
 
-Notably, the ZMQChannelsHandler has access to a kernel_client field (the kernel_client is created with `kernel_manager.get_kernel(self.kernel_id).client`).
+The kernel_manager we are referring to in Jupyter Server should not be confused with the kernel_manager of the Kernel Manager it self.
+To avoid confusion, we will name the Servers's one `mapping_kernel_manager` in the next sections.
+
+Notably, the ZMQChannelsHandler has access to a kernel_client field (the kernel_client is created with `mapping_kernel_manager.get_kernel(self.kernel_id).client`).
```
Comment (kevin-bates) [inline docs/server.rst]: The client isn't created by accessing the `client` field, but during the launch sequence (see KernelInterface constructor - which is invoked via the launch class method).  That field is purely an accessor.  I would just say something like _"...ZMQChannelsHandler has access to the kernel's client interface via its kernel_client property"_.

Comment (echarles) [reply docs/server.rst]: yeah I realized that while drawing the sequence diagram but have not update the text. Now it is done 👍

**Comment by echarles** [comment]
@kevin-bates has eagle eyes. Obviously my last push is a copy past of his comments. Thx again!

File: docs/server.rst [hunk 7]
```diff
@@ -0,0 +1,54 @@
+.. _server:
+
+=======================
+Use with Jupyter Server
+=======================
+
+This page describes the way `Jupyter Server <https://github.com/jupyter/jupyter_server>`_ uses the Kernel Management module.
+
+.. image:: ./images/kernel_mgmt_server.svg
```
Comment (takluyver) [inline docs/server.rst]: Minor nitpick in this image: the connection between Jupyter server and JKM is labelled "Jupyter Protocol (ZeroMQ)". This isn't quite right - Jupyter Server is using JKM through Python APIs (classes & functions), and through that talks the Jupyter protocol to the kernels themselves.

File: docs/server.rst [hunk 8]
```diff
@@ -0,0 +1,54 @@
+.. _server:
+
+=======================
+Use with Jupyter Server
+=======================
+
+This page describes the way `Jupyter Server <https://github.com/jupyter/jupyter_server>`_ uses the Kernel Management module.
```
Comment (takluyver) [inline docs/server.rst]: Ultimately this page probably belongs in the Jupyter Server docs, but if it makes sense to put it here at the moment, I don't have a problem with that.

File: docs/standalone.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (takluyver) [reply docs/standalone.rst]: I may rework this more at a later point. It is meant to be a public Python API as well as a component for Jupyter server.

File: docs/server.rst (hunk 7 above)
Comment (echarles) [reply docs/server.rst]: @takluyver great catch. I have updated the diagram to reflect this. There is still room for improvements on that diagram, but I think it is good-enough for now and at least does not show wrong info anymore.

File: docs/server.rst (hunk 8 above)
Comment (echarles) [reply docs/server.rst]: I will work next week with @Zsailer on the jupyter server doc. If this is merged before, we can reshuffle the details and also do a second pass on the kernel mgmt docs.

**Comment by echarles** [comment]
I have added the `hl` api in the docs, but the python doc is not generated although the `.. currentmodule:: jupyter_kernel_mgmt.hl` is defined.

```
WARNING: don't know which module to import for autodocumenting 'start_kernel_async' (try placing a "module" or "currentmodule" directive in the document, or giving an explicit module name)
```

**Comment by echarles** [comment]
- API page generation works - Needed to install jupyter_client as we have some rest reference here and there. Added jupyter_client in the requirements.yml
- Added a dedicate section for HL API

This is ready for review.

File: docs/api/hl.rst [hunk 9]
```diff
@@ -0,0 +1,12 @@
+.. _hl_api:
+
+High Level API
+==============
+
+.. automethod:: jupyter_kernel_mgmt.hl.start_kernel_async
```
Comment (takluyver) [inline docs/api/hl.rst]: Nitpick: I'd like to document these without referring to the `hl` module, e.g.  `jupyter_kernel_mgmt.start_kernel_async`. That's effectively an implementation detail.

File: docs/kernel_hl_api.rst [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+:class:`start_kernel_async <jupyter_kernel_mgmt.hl.start_kernel_async>`: Start a kernel by kernel type name. Returns a manager and a async client.
+
+:class:`start_kernel_async <jupyter_kernel_mgmt.hl.run_kernel_async>`: Context manager to run a kernel by kernel type name. Returns an async client.
+
+If the async is not supported on your platform, you may use the following blocking methods.
+
+:class:`start_kernel_blocking <jupyter_kernel_mgmt.hl.start_kernel_blocking>`: Start a kernel by kernel type name. Returns a manager and a blocking client.
+
+:class:`run_kernel_blocking <jupyter_kernel_mgmt.hl.run_kernel_blocking>`: Context manager to run a kernel by kernel type name. Returns a blocking client.
```
Comment (takluyver) [inline docs/kernel_hl_api.rst]: This is really just duplicating the API docs, so I'd rather leave it out.

File: docs/api/hl.rst (hunk 9 above)
Comment (echarles) [reply docs/api/hl.rst]: I have tried to define a currentmodule like in the othe rst files but I receive for the methods the following warning and at the end, the methods are not generated.

WARNING: don't know which module to import for autodocumenting 'start_kernel_async' (try placing a "module" or "currentmodule" directive in the document, or giving an explicit module name)

File: docs/kernel_hl_api.rst (hunk 10 above)
Comment (echarles) [reply docs/kernel_hl_api.rst]: I have now removed that file.

File: docs/api/hl.rst (hunk 9 above)
Comment (echarles) [reply docs/api/hl.rst]: Tried with other configuration, but not sure how sphynx handles the top level methods instead of the class methods.

**Pull request merged**